test/e2e/kcp_remediations.go

```diff
@@ -0,0 +1,710 @@
+	"sigs.k8s.io/cluster-api/test/framework"
+	"sigs.k8s.io/cluster-api/test/framework/clusterctl"
+	"sigs.k8s.io/cluster-api/util"
+)
+
+const (
+	configMapName    = "mhc-test"
+	configMapDataKey = "signal"
+
+	failLabelValue = "fail"
+)
+
+// KCPRemediationSpecInput is the input for KCPRemediationSpec.
+type KCPRemediationSpecInput struct {
+	// This spec requires following intervals to be defined in order to work:
+	// - wait-cluster, used when waiting for the cluster infrastructure to be provisioned.
+	// - wait-machines, used when waiting for an old machine to be remediated and a new one provisioned.
+	// - check-machines-stable, used when checking that the current list of machines in stable.
+	// - wait-machine-provisioned, used when waiting for a machine to be provisioned after unblocking bootstrap.
+	E2EConfig             *clusterctl.E2EConfig
+	ClusterctlConfigPath  string
+	BootstrapClusterProxy framework.ClusterProxy
+	ArtifactFolder        string
+	SkipCleanup           bool
+
+	// Flavor, if specified, must refer to a template that has a MachineHealthCheck
+	// - 3 node CP, no workers
+	// - Control plane machines having a pre-kubeadm command that queries for a well-known ConfigMap on the management cluster,
+	//   holding up bootstrap until a signal is passed via the config map.
+	//   NOTE: In order for this to work communications from workload cluster to management cluster must be enabled.
+	// - An MHC targeting control plane machines with the mhc-test=fail labels and
+	//     nodeStartupTimeout: 30s
+	// 	   unhealthyConditions:
+	//     - type: e2e.remediation.condition
+	//       status: "False"
+	// 	     timeout: 10s
+	// If not specified, "kcp-remediation" is used.
+	Flavor *string
+}
+
+// KCPRemediationSpec implements a test that verifies that Machines are remediated by MHC during unhealthy conditions.
+func KCPRemediationSpec(ctx context.Context, inputGetter func() KCPRemediationSpecInput) {
+	var (
+		specName         = "kcp-remediation"
+		input            KCPRemediationSpecInput
+		namespace        *corev1.Namespace
+		cancelWatches    context.CancelFunc
+		clusterResources *clusterctl.ApplyClusterTemplateAndWaitResult
+	)
+
+	BeforeEach(func() {
+		Expect(ctx).NotTo(BeNil(), "ctx is required for %s spec", specName)
+		input = inputGetter()
+		Expect(input.E2EConfig).ToNot(BeNil(), "Invalid argument. input.E2EConfig can't be nil when calling %s spec", specName)
+		Expect(input.ClusterctlConfigPath).To(BeAnExistingFile(), "Invalid argument. input.ClusterctlConfigPath must be an existing file when calling %s spec", specName)
+		Expect(input.BootstrapClusterProxy).ToNot(BeNil(), "Invalid argument. input.BootstrapClusterProxy can't be nil when calling %s spec", specName)
+		Expect(os.MkdirAll(input.ArtifactFolder, 0750)).To(Succeed(), "Invalid argument. input.ArtifactFolder can't be created for %s spec", specName)
+		Expect(input.E2EConfig.Variables).To(HaveKey(KubernetesVersion))
```

Comment:
should we validate that `input.Flavor` is also not nil up here?

Reply:
Please comment the underlying PR here: https://github.com/kubernetes-sigs/cluster-api/pull/7963

I just opened a PR to try to fix up the e2e test which is broken on 7963

Reply:
ACK